File: node/chapt2/chapt2.4/chapt2_ws/src/demo_cpp_pkg/src/person_node.cpp
#include "rclcpp/rclcpp.hpp"

// 定义 一个类class
class PersonNode : public rclcpp::Node
{ // 继承 使用 public；继承 rclcpp包 下的 Node类

    // 声明 成员数据（属性）【私有 成员】
private: // 成员声明： 数据类型  成员名      ； 加入 下划线 表示为 私有成员
    std::string name_;
    int age_;

    // 声明 成员函数（方法、构造函数）【共有 函数，可被调用】
public:
    // 设置成员数据，通过初始化列表(onst std::string &node_name)，来显式地调用 父类Node()的 成员数据/属性
    PersonNode(const std::string &node_name, const std::string &name, const int age)
        : Node(node_name) /* 调用 父类 的 构造函数*/ // 将PersonNode获取的node_name，传入 父类Node中
    {
        // 对 成员数据（属性） 进行 赋值: 使用 this -> 属性 = 形参      [this 表示 这个 类]
        this->name_ = name;
        this->age_ = age;
    }

    // 声明 方法
    void eat(const std::string &food_name)
    {

        // 使用 node父类的方法: 使用 this 说明 get_logger()是 继承到该类中的
        RCLCPP_INFO(this->get_logger(),"我是%s,%d岁，爱吃%s", this->name_.c_str(), // name_.c_str() 将 char类型的name 转换成 str类型
                    this->age_, food_name.c_str());                        // name 与 age 都为 私有属性，所以用 this 来调用

    }; // 短函数，可在 类中定义（结尾需加分号；）

}; // 必须有 分号；


int main(int argc,char **argv){

    // 初始化 节点
    rclcpp::init(argc,argv);
    
    // 创建对象:实例化 类
    // make_shared 将 node 变为一个 PersonNode类的指针  【智能指针】
    auto node = std::make_shared<PersonNode>("person_node","zyx",18);     // 将 PersonNode需要的参数(包括父类需要的)，都加上
    // 打印日志
    RCLCPP_INFO(node->get_logger(),"hello c++ node");
    
    // 运行 eat
    node -> eat("鱼香肉丝") ;               // 使用 -> 来 调用 成员函数（方法）

    // 运行 节点
    rclcpp::spin(node);
    // 清理内存
    rclcpp::shutdown();
    
    return 0;
}

File: node/chapt2/chapt2.4/chapt2_ws/src/demo_python_pkg/demo_python_pkg/person_node.py
import rclpy                    # 导入 rclpy库（功能包）
from rclpy.node import Node     # 从 rclpy功能包中的node节点文件 中， 引入 Node类（用于继承）


# 定义 一个类 的声明        【第一个参数一定要是 self】
class PersonNode(Node):                             # 括号中，加入待继承的 Node类
    # 定义 init方法：表示 定义属性（init函数 会在实例化 时，自动被调用）
    def __init__(self,node_name:str,name_value:str,age_value:int) -> None:     # self 是 固定存在的,表示 类自己；name_value 是 str类型（为一个 形参，需要实例化）
        
        # 引入 父类 的init属性
        super().__init__(node_name)

        # self.属性名: 表示 类的属性（只在 类中调用 属性时，用 self.）
        self.name = name_value                  # 表示：将name_value 传入的值，添加至 类的 name属性中，
        self.age = age_value                    #      定义 类 的一个属性的名字为 age，并将 age_value的值 传入

    # 定义 方法：（函数）
    def eat(self,food_name:str):        # 调用时，需要 输入一个变量
        # f 表示：python中的 格式化字符串。使用{ }可以调用 值
        # print(f"{self.name}今天过了自己的{self.age}岁的生日，并且吃了{food_name}")

        # 调用 父类的方法 来 打印字符（使用 self）
        # 因为 父类的方法 已经继承到了该类中，所以直接使用self即可调用该方法                            
        self.get_logger().info(f"{self.name}今天过了自己的{self.age}岁的生日，并且吃了{food_name}")                   


def main():
    rclpy.init()        # 初始化节点,分配资源
    
    # 实例化：创建对象node
    node = PersonNode('jiedian','zyx',25)
    # 使用对象，调用方法
    node.eat('蛋糕')

    rclpy.spin(node)    # 运行节点对象，并收集事件与执行
    rclpy.shutdown()    # 关闭节点，并清理内存
